### PR TITLE
Inject challenge

### PR DESCRIPTION
**Pull request by joejknowles**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,11 @@
+[![Build Status](https://travis-ci.org/joejknowles/inject-challenge.svg)](https://travis-ci.org/joejknowles/inject-challenge)
 Inject Challenge (Rubyist Badge)
 ================
 
+Approach
+--------
+I've created two methods (my_inject and my_inject_rec) which recreate inject's functionality. The first works iteratively and the second recursively.
```
Comment (sjmog) [inline README.md:7]: Lovely :+1:

File: lib/my_inject.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+class Array
+  def my_inject(*args)
+    start, accumulator = start_and_accumulator(args)
```
Comment (sjmog) [inline lib/my_inject.rb:3]: Beautiful stuff here

File: lib/my_inject.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    if block_given?
+      (start...length).each do |num|
```
Comment (sjmog) [inline lib/my_inject.rb:5]: Great use of implicit `self`

File: lib/my_inject.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        accumulator = yield(accumulator, self[num])
+      end
+    else
+      accumulator = run_symbol(start, accumulator, args[-1])
+    end
+    accumulator
+  end
+
+  def start_and_accumulator(args)
```
Comment (sjmog) [inline lib/my_inject.rb:14]: I might call this `generate_start_and_accumulator` for a tiny bit more clarity

File: lib/my_inject.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    if block_given?
+      (start...length).each do |num|
+        accumulator = yield(accumulator, self[num])
+      end
+    else
+      accumulator = run_symbol(start, accumulator, args[-1])
+    end
+    accumulator
+  end
+
```
Comment (sjmog) [inline lib/my_inject.rb:13]: Should these be `private`?

File: lib/my_inject.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      accumulator = args[0]
+    end
+    [start, accumulator]
+  end
+
+  def run_symbol(start, accumulator, symbol)
+    (start...length).each do |num|
+      accumulator = accumulator.send(symbol, self[num])
+    end
+    accumulator
```
Comment (sjmog) [inline lib/my_inject.rb:28]: Great terseness. Love this :+1:

File: lib/my_inject_rec.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  def my_inject_rec(*args, &block)
+    start, accumulator = start_and_accumulator_rec(args)
+    if block
+      accumulator = run_block_rec(start, accumulator, block)
+    else
+      accumulator = run_symbol_rec(start, accumulator, args[-1])
+    end
+    accumulator
+  end
+
```
Comment (sjmog) [inline lib/my_inject_rec.rb:11]: See before for use of `private` - at the moment any object can access `run_block_rec()` etc

File: lib/my_inject_rec.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  def run_block_rec(start, accumulator, block)
+    accumulator = block.call(accumulator, self[start])
+    start += 1
+    accumulator = run_block_rec(start, accumulator, block) if start < length
+    accumulator
+  end
```
Comment (sjmog) [inline lib/my_inject_rec.rb:17]: Lovely :+1:

File: lib/my_inject_rec.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    end
+    [start, accumulator]
+  end
+
+  def run_symbol_rec(start, accumulator, symbol)
+    accumulator = accumulator.send(symbol, self[start])
+    start += 1
+    accumulator = run_symbol_rec(start, accumulator, symbol) if start < length
+    accumulator
+  end
```
Comment (sjmog) [inline lib/my_inject_rec.rb:34]: Great recursion. Have you tried benchmarking these to see which runs faster?

File: spec/array_rec_spec.rb [hunk 10]
```diff
@@ -0,0 +1,72 @@
+require 'my_inject_rec'
+
+describe Array do
+  context 'recursive #inject' do
```
Comment (sjmog) [inline spec/array_rec_spec.rb:4]: Lovely use of RSpec convention

File: spec/array_rec_spec.rb [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    it 'takes block to add numbers in an array' do
+      expect([1, 4, 5].inject { |sum, num| sum + num }).to eq 10
+      expect([1, 4, 5].my_inject_rec { |sum, num| sum + num }).to eq 10
+    end
+    it 'takes block to subtract numbers in an array' do
+      expect([1, 4, 5].inject { |sum, num| sum - num }).to eq(-8)
```
Comment (sjmog) [inline spec/array_rec_spec.rb:10]: Lots and lots of use of `[1, 4, 5]` - maybe it would have made sense to spit this out into something like

``` ruby
let(:test_arr) { [1, 4, 5] }
```

File: spec/array_rec_spec.rb [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    end
+    it 'takes symbol to add numbers in an array with starting number' do
+      expect([1, 4, 5].inject(5, :+)).to eq 15
+      expect([1, 4, 5].my_inject_rec(5, :+)).to eq 15
+    end
+    it 'takes symbol to multiply numbers in an array with starting number' do
+      expect([1, 4, 5].inject(2, :*)).to eq 40
+      expect([1, 4, 5].my_inject_rec(2, :*)).to eq 40
+    end
+    it 'can recreate map when start is an array' do
```
Comment (sjmog) [inline spec/array_rec_spec.rb:53]: Dude, this test is intense. Nice edge case spot :+1:

File: spec/array_rec_spec.rb [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+      expect([1, 4, 5].inject([]) do |array, num|
+        array << num * 2
+      end).to eq [2, 8, 10]
+      expect([1, 4, 5].my_inject_rec([]) do |array, num|
+        array << num * 2
+      end).to eq [2, 8, 10]
+    end
+    it 'can find the longest word' do
```
Comment (sjmog) [inline spec/array_rec_spec.rb:61]: Very nice use of an `inject` edge case :+1:

File: spec/array_spec.rb [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+      longest = %w( cat sheep bear ).inject do |memo, word|
+        memo.length > word.length ? memo : word
+      end
+      expect(longest).to eq 'sheep'
+      longest = %w( cat sheep bear ).my_inject do |memo, word|
+        memo.length > word.length ? memo : word
+      end
+      expect(longest).to eq 'sheep'
+    end
+  end
```
Comment (sjmog) [inline spec/array_spec.rb:71]: ...aand another awesome spec file. Nice :+1:

**Comment by sjmog** [comment]
Amazing stuff here, Joe. I'm going to show this off to the coaches, if that's OK with you :+1: 

Lovely, full spec files - though look to see that you can save yourself some repetition by using `let`.

Beautiful classes - but bear in mind that everything you wrote was in a public interface (i.e. any other object in a program could access the inner workings directly). You're better off using `private` to protect the inner workings of objects.

Huge plus points for doing it both recursively and using iterators. You win!

**Comment by sjmog** [comment]
:checkered_flag:
